src/audio/aria/aria.c

```diff
@@ -0,0 +1,348 @@
+	cd->gain_state_ = 0;
+
+	return 0;
+}
+
+int aria_algo_buffer_data(struct comp_dev *dev, int32_t *__restrict data, size_t size)
+{
+	struct aria_data *cd = comp_get_drvdata(dev);
+	size_t min_buff = MIN(cd->buff_size_ - cd->buff_pos_ - cd->offset_, size);
+	int ret;
+
+	ret = memcpy_s(&cd->data_[cd->buff_pos_], cd->buff_size_ * sizeof(int32_t),
+		       data, min_buff * sizeof(int32_t));
+	if (ret < 0)
+		return ret;
+	ret = memcpy_s(cd->data_, cd->buff_size_ * sizeof(int32_t),
+		       &data[min_buff], (size - min_buff) * sizeof(int32_t));
+	if (ret < 0)
+		return ret;
+	cd->buff_pos_ = (cd->buff_pos_ + size) % cd->buff_size_;
+
+	return 0;
+}
+
+int aria_process_data(struct comp_dev *dev,
+		      int32_t *__restrict dst, size_t dst_size,
+		      int32_t *__restrict src, size_t src_size)
+{
+	struct aria_data *cd = comp_get_drvdata(dev);
+	size_t min_buff;
+	int ret;
+
+	if (cd->att_) {
+		aria_algo_calc_gain(dev, (cd->gain_state_ + 1) % ARIA_MAX_GAIN_STATES,
+				    src, src_size);
+		aria_algo_get_data(dev, dst, dst_size);
+	} else {
+	/* bypass processing gets unprocessed data from buffer */
+		min_buff = MIN(cd->buff_size_ - cd->buff_pos_ - cd->offset_, dst_size);
+		ret = memcpy_s(dst, cd->buff_size_ * sizeof(int32_t),
+			       &cd->data_[cd->buff_pos_ + cd->offset_],
+			       min_buff * sizeof(int32_t));
+		if (ret < 0)
+			return ret;
+		ret = memcpy_s(&dst[min_buff], cd->buff_size_ * sizeof(int32_t),
+			       cd->data_, (dst_size - min_buff) * sizeof(int32_t));
+		if (ret < 0)
+			return ret;
+	}
+	ret = aria_algo_buffer_data(dev, src, src_size);
+	return ret;
+}
+
+static int init_aria(struct comp_dev *dev, struct comp_ipc_config *config,
+		     void *spec)
+{
+	struct ipc4_aria_module_cfg *aria = spec;
+	struct aria_data *cd;
+	int ret;
+
+	dev->ipc_config = *config;
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	dcache_invalidate_region(spec, sizeof(*aria));
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		comp_free(dev);
+		return -ENOMEM;
+	}
+
+	/* Copy received data format to local structures */
+	ret = memcpy_s(&cd->base, sizeof(struct ipc4_base_module_cfg),
+		       &aria->base_cfg, sizeof(struct ipc4_base_module_cfg));
+	if (ret < 0)
+		return ret;
+
+	const size_t ibs = cd->base.ibs;
+	const size_t chc = cd->base.audio_fmt.channels_count;
+	const size_t sgs = (cd->base.audio_fmt.valid_bit_depth >> 3) * chc;
+	const size_t sgc = ibs / sgs;
+	const size_t req_mem = GetRequiredMemory(chc, sgc);
+	size_t att = aria->attenuation_;
+
+	if (aria->attenuation_ > ARIA_MAX_ATT) {
+		comp_warn(dev, "init_aria(): Attenuation value %d must not be greater than %d",
+			  att, ARIA_MAX_ATT);
+		att = ARIA_MAX_ATT;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	void *buf = rballoc(0, SOF_MEM_CAPS_RAM, req_mem);
+	void *buf_in = rballoc(0, SOF_MEM_CAPS_RAM, req_mem);
+	void *buf_out = rballoc(0, SOF_MEM_CAPS_RAM, req_mem);
+
+	if (!buf || !buf_in || !buf_out) {
```

Comment:
This leaks if one allocation fails.

Reply:
this is still leaking memory? we should free the successful allocation in case of failure